File: cpp/problems/leetcode/0004_median_of_two_sorted_arrays/solution.cpp
// LeetCode 4
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        vector<int> *A = &nums1, *B = &nums2;
        if (A->size() > B->size()) swap(A, B);
        int m = (int)A->size(), n = (int)B->size();
        int lo = 0, hi = m, half = (m + n + 1) / 2;
        while (lo <= hi) {
            int i = (lo + hi) / 2;
            int j = half - i;
            long long aL = i ? (*A)[i - 1] : LLONG_MIN / 4;
            long long aR = i < m ? (*A)[i] : LLONG_MAX / 4;
            long long bL = j ? (*B)[j - 1] : LLONG_MIN / 4;
            long long bR = j < n ? (*B)[j] : LLONG_MAX / 4;
            if (aL <= bR && bL <= aR) {
                if ((m + n) % 2) return (double)max(aL, bL);
                return (max(aL, bL) + min(aR, bR)) / 2.0;
            }
            if (aL > bR)
                hi = i - 1;
            else
                lo = i + 1;
        }
        return 0;
    }
};

int main() {
    vector<int> a{1, 3}, b{2};
    assert(Solution().findMedianSortedArrays(a, b) == 2.0);
    vector<int> c{1, 2}, d{3, 4};
    assert(Solution().findMedianSortedArrays(c, d) == 2.5);
    cout << "0004 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0004_median_of_two_sorted_arrays/solution.py
"""LeetCode 4: Median of Two Sorted Arrays."""

from __future__ import annotations


class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        a, b = nums1, nums2
        if len(a) > len(b):
            a, b = b, a
        m, n = len(a), len(b)
        lo, hi = 0, m
        half = (m + n + 1) // 2
        while lo <= hi:
            i = (lo + hi) // 2
            j = half - i
            a_left = a[i - 1] if i else -10**18
            a_right = a[i] if i < m else 10**18
            b_left = b[j - 1] if j else -10**18
            b_right = b[j] if j < n else 10**18
            if a_left <= b_right and b_left <= a_right:
                if (m + n) % 2 == 1:
                    return float(max(a_left, b_left))
                return (max(a_left, b_left) + min(a_right, b_right)) / 2.0
            if a_left > b_right:
                hi = i - 1
            else:
                lo = i + 1
        return 0.0


if __name__ == "__main__":
    assert Solution().findMedianSortedArrays([1, 3], [2]) == 2.0
    assert Solution().findMedianSortedArrays([1, 2], [3, 4]) == 2.5
    print("0004 OK")
